src/smtsolvers/TheoryInterpolator.h

```diff
@@ -3,7 +3,139 @@
+ *       then f(x) will also be stored as shared.
+ */
+class LocalTermColorInfo : public TermColorInfo {
+public:
+
+    template<typename TMap>
+    LocalTermColorInfo(TMap const & topLevelMap, Logic const & logic) {
+        termColors[logic.getTerm_true()] = I_AB;
+        termColors[logic.getTerm_false()] = I_AB;
+        computeColorsForAllSubterms(topLevelMap, logic);
+    }
+
+    icolor_t getColorFor(PTRef term) override {
+        auto it = termColors.find(term);
+        if (it == termColors.end()) {
+            throw OsmtInternalException("No color detected for term");
+        }
+        return it->second;
+    }
+
+private:
+    std::unordered_map<PTRef, icolor_t, PTRefHash> termColors;
+
+    template<typename TMap>
+    void computeColorsForAllSubterms(TMap const & topLevelColors, Logic const & logic) {
+        // MB: NOTE! If P(a) is A-local, but both symbols P and a are shared, than P(a) should be shared and not A-local
+        using entry_t = std::pair<const PTRef, icolor_t>;
+        auto colorUnion = [](icolor_t f, icolor_t s) { return static_cast<opensmt::icolor_t>(f | s); };
+        std::vector<entry_t> queue;
+        for (auto entry : topLevelColors) {
+            queue.push_back(entry);
+        }
+        std::unordered_map<SymRef, icolor_t, SymRefHash> symbolColors;
+        while (not queue.empty()) {
+            auto const & entry = queue.back();
+            icolor_t colorToAssign = entry.second;
+            PTRef term = entry.first;
+            queue.pop_back();
+            auto it = termColors.find(term);
+            if (it != termColors.end()) {
+                icolor_t assignedColor = it->second;
+                if (assignedColor == colorToAssign || assignedColor == I_AB) { // already processed, color does not change
+                    continue;
+                } else { // assigning new color
+                    assert(assignedColor == I_A or assignedColor == I_B);
+                    colorToAssign = colorUnion(colorToAssign, assignedColor);
+                    assert(colorToAssign == I_AB);
+                }
+            }
+            // if we reach here, we need to propagate colorToAssign to the whole term subtree of `term`
+            termColors[term] = colorToAssign;
+            for (PTRef child : logic.getPterm(term)) {
+                queue.emplace_back(child, colorToAssign);
+            }
+            // add symbol information
+            auto insertRes = symbolColors.insert(std::make_pair(logic.getSymRef(term), colorToAssign));
+            if (not insertRes.second) { // there was entry for this symbol already
+                auto entryIt = insertRes.first;
+                entryIt->second = colorUnion(entryIt->second, colorToAssign);
+            }
+        }
+        // Make sure complex terms have correct color assigned
+        vec<PTRef> terms;
+        for (auto const & entry : termColors) {
```

Comment:
Consider having instead `for (auto const entry : termColors) {` ?

Reply:
Replaced